Add 'biothings-admin dataplugin create' command

remoteeng00 opened the pull request:
ref https://github.com/biothings/biothings.api/issues/250

- Installing: please run  `pip install -e ".[hub]"`
- Using: 
   - Create a single dumper plugin: `biothings-admin dataplugin create --name test --plugin-type single`
   - Create a multiple dumpers plugin: `biothings-admin dataplugin create --name test --plugin-type multiple`
   - Add `--parallelizer` if you want to create a plugin with parallelizer setup
   - Create plugin in step by step: `biothings-admin dataplugin create` 
   - See help: `biothings-admin dataplugin create --help`

**Update**
We have two set of commands that alternative to two mode of testing data plugins:
  - `biothings-admin standalone-dataplugin [create | dump_and_upload | inspect | serve ]` - running out side of a data plugin folder, test locally without the need of docker, mongodb, dedicated hub server
  - `biothings-admin dataplugin [dump | upload | inspect | serve | list | clean]` - running under a data plugin folder, direct testing under a data plugin folder 

More info:
  - `biothings-admin dataplugin --help`
  - `biothings-admin standalone-dataplugin --help`

remoteeng00 committed:
Add 'biothings-admin dataplugin create' command

newgene commented on biothings/management/dataplugin.py:
```diff
@@ -0,0 +1,44 @@
+from typing import Optional
+
+import typer
+from jinja2 import Environment, FileSystemLoader
```
@remoteeng00 do you think tornado's templating feature is good enough here?

https://www.tornadoweb.org/en/stable/template.html

if so, we can remove the Jinja2 dependency here.

remoteeng00 replied:
yeap, it's updated.

newgene commented on biothings/management/dataplugin.py:
```diff
@@ -0,0 +1,44 @@
+    biothing_source_dir = pathlib.Path(__file__).parent.parent.resolve()
+    template_dir = os.path.join(biothing_source_dir, 'hub', 'dataplugin', 'templates')
+    plugin_dir = os.path.join(workspace_dir, name)
+    assert os.path.isdir(template_dir), "Template folder does not exists"
```
use of assert here for error msg is not very user friendly. We should change it to normal way to check and output error msg.

newgene commented on biothings/management/dataplugin.py:
```diff
@@ -0,0 +1,44 @@
+    assert not os.path.isdir(
+        plugin_dir
+    ), "Data plugin with the same name is already exists, please remove it before create"
+    cmd = ["cp", "-r", template_dir, plugin_dir]
```
@remoteeng00 `cp` won't work for windows. For creating a data plugin, we should make it work under Windows too.

newgene commented on setup.py:
```diff
@@ -79,6 +79,8 @@ def read(fname):
     'requests-aws4auth',  # aws s3 auth requests for autohub
     'networkx>=2.1,<2.6',  # datatransform
     'biothings_client>=0.2.6',  # datatransform (api client)
+    'typer[all]',  # Lib for building CLI applications
```
Tornado also has `tornado.options` module: https://www.tornadoweb.org/en/stable/options.html.

But I think `typer` is very nice. You can evaluate `tornado.options` to see if it's good enough. I am ok to use `typer` if necessary.

remoteeng00 replied:
I prefer to use `typer`, it help us easily parse arguments and define new command.

newgene commented on biothings/hub/dataplugin/templates/parser.py:
```diff
@@ -0,0 +1,14 @@
+    pass
+
+
+def custom_mapping(cls):
```
I believe we typically put mapping in a separate `mapping.py` file.

newgene commented on biothings/hub/dataplugin/templates/parallelizer.py:
```diff
@@ -0,0 +1,5 @@
+def custom_jobs(self):
```
`parallelizer` is a bit complicated for most of data plugin developers, so I would not include it by default, but can be included when an optional cli argument is passed.

newgene commented on biothings/hub/dataplugin/templates/manifest.yaml.j2:
```diff
@@ -0,0 +1,34 @@
+---
+version: '0.3'
+__metadata__: # Optiona.
```
just a typo: `Optiona` --> `Optional`

newgene commented on biothings/management/dataplugin.py:
```diff
@@ -0,0 +1,51 @@
+        help="Data plugin name",
+        prompt="What's your data plugin name?",
+    ),
+    plugin_type: Optional[str] = typer.Option(  # NOQA: B008
```
90% cases are single uploader, so no need to make it a required question, an optional `--multi-uploaders` to turn it on is good enough.

newgene commented on biothings/hub/dataplugin/templates/manifest.yaml.tpl:
```diff
@@ -0,0 +1,35 @@
+__metadata__: # Optional.
+  license_url: https://example.com/  # Optional. Put your license url here
+  licence: ABCXYZ # Optional. Your license name
+  url: https://example.com/  # Your site url
```
@remoteeng00 let's add an optional `description` field under `__metadata__`. It can be used to provide an optional description text about this plugin

remoteeng00 committed:
[WIP] Fix: get_src_conn in the hub_db utils, make sqlite3 utils works like the mongo utils;  Add: biothings-admin dataplugin test_dump_and_upload, test_inspect;

remoteeng00 committed:
Add new CLI params; Fix data plugin template; Fix sqlite3 utils in order to run bulk_write from MergerStorage & IgnoreDuplicatedStorage storage

remoteeng00 committed:
Fix: parsing source table name & source full name from CLI option

remoteeng00 committed:
Add command `dataplugin server` to start a simple API

remoteeng00 committed:
Support data plugin API filters, add sample config.py, raise exception when run inspect CLI with missing params

sengineer0 committed:
refactor uitls.diff: split into 2 files, one not require to setup Biothings Hub, the other still requires

remoteeng00 committed:
Move the storage module and upload_worker to the utils pakage in order to run it without hub

remoteeng00 committed:
Add standalone-dataplugin CLI commands (dump, upload, inspect, serve)

remoteeng00 committed:
Fix: support both manifest.json and manifest.yaml

remoteeng00 committed:
Reformat help text, add option about verbosity logging

remoteeng00 committed:
Fix import load data module from data plugin path

remoteeng00 committed:
Rename module, update CLI commands; Fix import config.py in current folder

newgene commented on biothings/hub/__init__.py:
```diff
@@ -29,6 +29,7 @@
 def _config_for_app(config_mod=None):
 
     if not config_mod:
+        sys.path.insert(0, ".")  # import config.py in current folder
```
probably should check if the current folder is already in the `sys.path` to avoid adding it repeatedly. @remoteeng00

remoteeng00 committed:
Rebase from master; Update the config loading following new DummyConfig class

remoteeng00 committed:
Fix minor issue

sengineer0 committed:
Implement flatten and validation data inspection

sengineer0 committed:
Implement api endpoint to get flattened inspection data

remoteeng00 committed:
Add CLI commands: 'clean', 'list'; Add 'host' option on 'serve' command; Intergrate with the flatten feature on 'inspect' command

remoteeng00 committed:
Add missing requirement

remoteeng00 committed:
Add storage module for backward compatibl

remoteeng00 committed:
Use warning module to display message

remoteeng00 committed:
Fix 'clean' command help text;
Update 'inspect' and 'serve' command output;
Update query in 'serve' command;
Fix field validations;
Support 'limit' parameter in 'upload' command

sengineer0 committed:
Refactor inspection data feature
- add warning code
- refactor validation logic for easier adding more rules later

remoteeng00 committed:
Remove [] in the filter params

remoteeng00 committed:
Update help text; Update inspect command with warning codes; Update serve command: new response, fix query AND, support ':' in query value

remoteeng00 committed:
Support wilcard searching

remoteeng00 committed:
Update UI using rich; Update some behaviors: default will list all, and inspect all

remoteeng00 committed:
Merge branch 'master' into add-biothings-admin-cli

remoteeng00 committed:
Add useful behaviors, beautify messages

remoteeng00 commented:
Doc for **biothings-admin dataplugin-hub** commands:
 Usage: biothings-admin dataplugin-hub   [OPTIONS] COMMAND [ARGS]

 Bootstrap your data plugin as the Hub does but in a synchronized. Create your new data plugin with the standard file structure. Dumping, uploading, and inspecting your data in synchronize. Serving your data as a web
 service for making simple queries * Running this command outside of your data plugin

Commands:
- **create**: Create a new data plugin                                                                                                          
-  **dump_and_upload**: Download data source to local folder then convert to Json document and upload to the source database  
- **inspect**: Giving detailed information about the structure of documents coming from the parser                                               
- **serve**: Run the simple API server for serving documents from above 'dump_and_upload' command

remoteeng00 committed:
Fix show mapping when inspect with mapping mode

remoteeng00 committed:
Add list/clean command in dataplugin-hub

remoteeng00 committed:
Update CLI displaying; Add -o option; Fix mapping does not show in some cases

remoteeng00 committed:
Fix exception when missing stats in the mode param of the inspect cmd

remoteeng00 committed:
Try to import config.py locally when runing CLI
